src/ctap/pin_protocol.rs

```diff
@@ -0,0 +1,460 @@
+    pub fn new(rng: &mut impl Rng256) -> PinProtocol {
+        let key_agreement_key = crypto::ecdh::SecKey::gensk(rng);
+        let pin_uv_auth_token = rng.gen_uniform_u8x32();
+        PinProtocol {
+            key_agreement_key,
+            pin_uv_auth_token,
+        }
+    }
+
+    /// Generates a fresh public key.
+    pub fn regenerate(&mut self, rng: &mut impl Rng256) {
+        self.key_agreement_key = crypto::ecdh::SecKey::gensk(rng);
+    }
+
+    /// Generates a fresh pinUvAuthToken.
+    pub fn reset_pin_uv_auth_token(&mut self, rng: &mut impl Rng256) {
+        self.pin_uv_auth_token = rng.gen_uniform_u8x32();
+    }
+
+    /// Returns the authenticator’s public key as a CoseKey structure.
+    pub fn get_public_key(&self) -> CoseKey {
+        CoseKey::from(self.key_agreement_key.genpk())
+    }
+
+    /// Processes the peer's encapsulated CoseKey and returns the shared secret.
+    pub fn decapsulate(
+        &self,
+        peer_cose_key: CoseKey,
+        pin_uv_auth_protocol: u64,
+    ) -> Result<Box<dyn SharedSecret>, Ctap2StatusCode> {
+        let pk: crypto::ecdh::PubKey = CoseKey::try_into(peer_cose_key)?;
+        let handshake = self.key_agreement_key.exchange_x(&pk);
+        match pin_uv_auth_protocol {
+            1 => Ok(Box::new(SharedSecretV1::new(handshake))),
+            2 => Ok(Box::new(SharedSecretV2::new(handshake))),
+            _ => Err(Ctap2StatusCode::CTAP1_ERR_INVALID_PARAMETER),
+        }
+    }
+
+    /// Getter for pinUvAuthToken.
+    pub fn get_pin_uv_auth_token(&self) -> &[u8; PIN_TOKEN_LENGTH] {
+        &self.pin_uv_auth_token
+    }
+
+    /// This is used for debugging to inject key material.
+    #[cfg(test)]
+    pub fn new_test(
+        key_agreement_key: crypto::ecdh::SecKey,
+        pin_uv_auth_token: [u8; PIN_TOKEN_LENGTH],
+    ) -> PinProtocol {
+        PinProtocol {
+            key_agreement_key,
+            pin_uv_auth_token,
+        }
+    }
+}
+
+/// Verifies the pinUvAuthToken for the given PIN protocol.
+pub fn verify_pin_uv_auth_token(
+    token: &[u8; PIN_TOKEN_LENGTH],
+    message: &[u8],
+    signature: &[u8],
+    pin_uv_auth_protocol: u64,
+) -> Result<(), Ctap2StatusCode> {
+    match pin_uv_auth_protocol {
+        1 => verify_v1(token, message, signature),
+        2 => verify_v2(token, message, signature),
+        _ => Err(Ctap2StatusCode::CTAP1_ERR_INVALID_PARAMETER),
+    }
+}
+
+pub trait SharedSecret {
+    /// Returns the encrypted plaintext.
+    fn encrypt(
+        &mut self,
+        rng: &mut dyn Rng256,
+        plaintext: &[u8],
+    ) -> Result<Vec<u8>, Ctap2StatusCode>;
+
+    /// Returns the decrypted ciphertext.
+    fn decrypt(&mut self, ciphertext: &[u8]) -> Result<Vec<u8>, Ctap2StatusCode>;
+
+    /// Verifies that the signature is a valid MAC for the given message.
+    fn verify(&mut self, message: &[u8], signature: &[u8]) -> Result<(), Ctap2StatusCode>;
+}
+
+fn encrypt(
```

Comment:
`aes256_cbc_encrypt` would be more descriptive here.

Reply:
Renamed.